File: src/ts/terrain/landscape/continentLayer.ts
import { simpleElevationFunction } from "./elevationFunction";
import { LVector3 } from "../../utils/algebra";
import { simplex401, simplex411 } from "../../utils/simplex";
import { simpleFractalLayer3D } from "./simpleFractalLayer3D";

export function continentLayer(frequency: number, nbOctaves: number): simpleElevationFunction {
    return simpleFractalLayer3D(frequency, nbOctaves, 2.0, 2.0, 0.8, 0.0, (coords: LVector3, seed: number, gradient: LVector3) => {
        const warpedCoords = coords.clone();
        const warpingStrength = 0.6;
        warpedCoords.x += warpingStrength * simplex411(warpedCoords, seed);
        warpedCoords.y += warpingStrength * simplex411(warpedCoords.add(new LVector3(13, 37, -67)), seed);
        warpedCoords.z += warpingStrength * simplex411(warpedCoords.add(new LVector3(-53, 19, -29)), seed);

        return simplex401(warpedCoords, seed, gradient);
    });
}

File: packages/terrain-generation/src/landscape/continent_layer.rs
use crate::utils::math::{multiply, smoothstep};
use crate::utils::simplex::simplex401;
use crate::utils::vector3::Vector3;

pub fn continent_layer(
    frequency: f32,
    continents_fragmentation: f32,
) -> impl Fn(&Vector3, f32, &mut Vector3) -> f32 {
    move |coords: &Vector3, seed: f32, gradient: &mut Vector3| {
        let decay = 2.0;
        let lacunarity = 2.0;
        let nb_octaves = 6;
        let mut noise_value = 0.0;
        let mut total_amplitude = 0.0;
        let mut local_gradient = Vector3::zero();
        for i in 0..nb_octaves {
            let local_frequency = frequency * f32::powi(lacunarity, i);
            let local_amplitude = 1.0 / f32::powi(decay, i);

            let mut local_elevation =
                simplex401(&(coords * local_frequency), seed, &mut local_gradient);
            local_gradient *= local_frequency;

            local_elevation = smoothstep(0.1, 1.0, local_elevation, &mut local_gradient);

            local_elevation *= local_amplitude;
            local_gradient *= local_amplitude;

            total_amplitude += local_amplitude;

            noise_value += local_elevation;
            *gradient += &local_gradient;
        }
        noise_value /= total_amplitude;
        *gradient /= total_amplitude;

        smoothstep(continents_fragmentation, 1.0, noise_value, gradient);

        let mut detail_gradient = Vector3::zero();
        let mut detail_noise = simplex401(&(coords * 10.0), seed, &mut detail_gradient);
        detail_gradient *= 10.0;

        detail_noise = multiply(detail_noise, noise_value, &mut detail_gradient, gradient);

        noise_value -= detail_noise * 0.3;
        *gradient -= &(detail_gradient * 0.3);

        noise_value
    }
}
